### PR TITLE
Nitro Service API definition ADR

### PR DESCRIPTION
**Pull request by mod**
# Nitro As a Service

Request for comments and approval.
Once the ADR is mature, our team will provide a starting API implementation in WS JSON with a migration of payment app outside the framework.

Naming suggestions for the Objective router:
- protocol_manager
- objective_manager
- app_manager

Message structure would only need to container Objectives array, but not payment messages,
Payment app, would use it's own network and transport to exchange vouchers.

### CODE REVIEW

File: .adr/0009-nitro-service.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+#### Response:
+`[type, msgid, error, result]`
+
+#### Event:
+`[type, msgid, method, params]`
+
+##### type
+The message type, must be the integer zero (0) for "Request" messages. 
+One (1) means that this message is the "Response" message.
+Two (2) means that this message is the "Notification" message.
```
Comment (lalexgap) [inline .adr/0009-nitro-service.md:53]: Is this similar to the JSON-RPC [notifications](https://www.jsonrpc.org/specification#notification) where "notifications" is just a request from a client where the client doesn't care about the response?

File: .adr/0009-nitro-service.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+##### msgid
+A 32-bit unsigned integer number. This number is used as a sequence number. The server's response to the "Request" will have the same msgid.
+
+Events streams may also relate to a subscription "Request".
```
Comment (lalexgap) [inline .adr/0009-nitro-service.md:58]: > Events streams may also relate to a subscription "Request".

Does this mean that the API would take subscription "requests" and return some kind of event stream?

**Review by lalexgap — COMMENTED** [review]
Thanks for putting this together, it's great to get feedback from those wanting to use our API!

I've added some questions/comments that would be great to get your feedback on.

**Comment by lckmidnight** [comment]
I would suggest to combine `type` and `method`.
In fact, a method will always be considred as a specific type of message (request/response/event)

In the examples:
`[0, 42, "create_objective", [protocol, peers, app_definition]]`
The create objective method would always be considered as a request inside the protocol

Same apply for responses and events

So i would suggest instead to used a single `type` field (or `method`), like these:
- create_objective (request)
- objective_created (response)
- objective_update (event)
- objective_completed (event)

It is also possible to find a conventions for method names, by naming them like: `create_objective_request` which would correspond with a `create_objective_response`

File: .adr/0009-nitro-service.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mod) [reply .adr/0009-nitro-service.md:53]: Event = notifications indeed.
there can be all kind of push notification events from nitro to clients, when protocol is successfully deployed and errors, or if someone calls a "defund" or  challenge occurs.

The transport must be persistent bi-directional TCP connection, WebSocket, HTTP2, gRPC, Nats (and other MQ)
would all work but not simple HTTP Rest.

I suggest we start with the following implementation:
- json-rpc
- Msgpack version with my protocol as describe (faster performance)
- protobuf

Transports:
- WebSocket
- Nats
- gRPC (it's HTTP2)

File: .adr/0009-nitro-service.md (hunk 2 above)
Comment (mod) [reply .adr/0009-nitro-service.md:58]: that's correct, the event stream id, match the request id.

File: .adr/0009-nitro-service.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+#### error
+If the method is executed correctly, this field is Nil. If the error occurred at the server-side, then this field is an arbitrary object which represents the error.
+
+#### result
+An arbitrary object, which represents the returned result of the function. If an error occurred, this field should be nil.
+
+### create_objective
+
+`[0, 42, "create_objective", [protocol, peers, app_definition]]`
```
[the remark this answers is not in the document]
Comment (mod) [reply .adr/0009-nitro-service.md:74]: Indeed, I need to address this well,
First I will make a gRPC version to illustrate well,
then maybe json-rpc

I suggest then:
`[0, 42, "execute_protocol", [protocol, peers, app_definition, paramStruct... ] ]`

File: .adr/0009-nitro-service.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+- app_definition: **String**, smart-contract address of the NitroApp.
+
+#### Response:
+
+`[1, 42,  nil, objective_id]`
+
+Params:
+- **objective_id**: **String**, unique identifier of the Objective request.
+
+### objective_created Event
```
[the remark this answers is not in the document]
Comment (mod) [reply .adr/0009-nitro-service.md:89]: that's what I meant by created, I will rename to "protocol_deployed"

File: .adr/0009-nitro-service.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+`[2, 42, "objective_created", [id, proof]]`
+
+Params:
+- id, **string**, Objective identifier
+### cancel_objective
```
[the remark this answers is not in the document]
Comment (mod) [reply .adr/0009-nitro-service.md:95]: Then I suggest we postpone this topic as it is not urgent, but if other methods come to your mind please suggest

File: .adr/0009-nitro-service.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mod) [reply .adr/0009-nitro-service.md:53]: @lalexgap check this out:
https://pkg.go.dev/github.com/filecoin-project/go-jsonrpc#section-readme

**Pull request closed without merging**